### PR TITLE
#482: Improve and complete JSON writing

### LINKED ISSUE #482: Improve and complete JSON writing
The goal is to be able to have the JSON outputted by LBAF (when using the internal data structures) to be identical to the input JSON.

Missing elements at this point are (non exhaustive):
- [x] Communications
- [x] subphases
- [x] index
- [x] collection_id
- [x] shared_node data

This issue should also deprecate the use of `-1` to read/step through all phases (only explicit enumerations or ranges should be used).

### PR DESCRIPTION
**Pull request by cwschilly**
Fixes #482

This PR also deprecates the `-1` option (to balance all phases). If no phase is specified, it defaults to balancing phase 0.

### CODE REVIEW

**Comment by cwschilly** [comment]
I've found a bug in the code--putting in draft until I fix it

File: src/lbaf/Model/lbsObject.py [hunk 1]
```diff
@@ -163,3 +166,11 @@ def set_communicator(self, c) -> None:
     def get_subphases(self) -> list:
         """Return subphases of this object."""
         return self.__subphases
+
+    def set_unused_params(self, unused_params: dict):
+        """Assign any extraneous parameters."""
+        self.__unused_params = unused_params
+
+    def get_unused_params(self) -> dict:
```
Comment (cwschilly) [inline src/lbaf/Model/lbsObject.py:174]: Due to the way we dynamically get all Object QOI, this will now list `unused_params` as a QOI

**Review by ppebay — APPROVED** [review]
Good , especially with unit test. Thanks

**Review by ppebay — APPROVED** [review]

**Comment by cwschilly** [comment]
I'm running into an issue with assigning communication dicts to the correct rank. Originally, I was pairing communications with ranks by checking to see which object the comm is from (`communications["from"]["id"]`). Then I would compare that id to the ids from `rank.get_object_ids()`. If they matched, I would write out that communication on that rank.

However, it seems like a lot of these communications are sent from `"home": 0, "id": 0`, but there isn't an object (`entity`, in the JSON) that has `id: 0` (meaning that there is never a match). Am I misunderstanding something?

**Review by ppebay — APPROVED** [review]
These are good changes.

**Review by pierrepebay — COMMENTED** [review]
Looks good otherwise. I have tested it as well.

File: src/lbaf/IO/lbsVTDataWriter.py [hunk 2]
```diff
@@ -62,41 +69,147 @@ def __create_tasks(self, rank_id, objects, migratable):
...
+
+        # Loop through any moved objects to find the correct rank
+        if self.__moved_comms:
+            for moved_dict in self.__moved_comms:
+                if moved_dict["from"]["id"] in rank_objects:
+                    communications.append(moved_dict)
+
+        return communications
+
+    def serialize(self, phases: dict) -> list:
```
[the remark this answers is not in the document]
Comment (cwschilly) [reply src/lbaf/IO/lbsVTDataWriter.py]: Fixed, thanks!

**Review by ppebay — APPROVED** [review]
Sorry @lifflander @pierrepebay I inadvertently requested a new review from everyone albeit only one from me was needed. Please ignore the new request.

**Pull request merged**